text/0000-result-expect.md

```diff
@@ -0,0 +1,40 @@
+- Feature Name: `result_expect`
+- Start Date: 2015-05-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add an `expect` method to the Result type, bounded to `E: Debug`
```

Comment:
According to [RFC 236](https://github.com/rust-lang/rfcs/blob/master/text/0236-error-conventions.md), which is focused on the standard library but is intended to be a guideline for Rust libraries in general, panics should only be used for bugs (not user facing) or when there is nothing else that you can meaningfully do (e.g. out of memory). This is why `unwrap` uses a `Debug` bound currently, AFAIK. Using `unwrap` or `expect` is basically an assertion that you will *never* get an `Err` value (unless there is a bug in your program, of course), and thus that a user will never see this error.

It is absolutely the case that whenever rustc panics, it is due to a bug. The code you linked should execute rarely or never in a release, but it exists in order to assist the user in filing a bug report when there *is* an error. There's no expectation that the user will be able to understand anything about the output, other than that they should send it to the Rust developers in a bug report.

My understanding is that, if you want a utility to be able to die in "normal" circumstances (i.e. not a bug in the program itself), it's preferable not to use an uncaught `panic`, but to print a message yourself and then call `std::process::exit`. We don't have any hard documentation that I know of about this, but I think that's in part because `std::process::exit` was only stabilized recently (a month or so ago, during the beta/release rush?).

If you want destructors to run, you do have to return up to main (or whatever level has the destructors you need) before exiting, which is an inconvenience in the current design. I think #1100 is intended to address this in part by allowing panics to be done more "quietly" and then caught at a higher level (but it is still very coarse).